### PR TITLE
dune 3.4.1 & Adaptation of ocaml/dune for loongson3 and riscv64

### PR DESCRIPTION
**Pull request by CamberLoid**
Topic Description
-----------------

The topic will upgrade dune to 3.4.1, and begin adaptation process about related package of riscv64 and loongson3 platform.

Package(s) Affected
-------------------

`ocaml-pp`: new, in bootstrap process
`findlib`: update to
`ocaml-csexp`: update to 1.5.1, in bootstrap process
`dune`: update to 3.4.1, bootstrap target

And a temporary minimal dune executable `minidune`, which is used only in bootstrap process and won't be included and shipped.

Security Update?
----------------

No

Build Order
-----------

amd64/arm64: `ocaml-pp ocaml-csexp findlib dune`
Not include loongson3 and riscv64 as they require additional operations to bootstrap.

Test Build(s) Done
------------------

**Primary Architectures**

- [x] AMD64 `amd64`   
- [x] AArch64 `arm64`
    

**Secondary Architectures**

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [x] Loongson 3 `loongson3`
- [ ] PowerPC 64-bit (Little Endian) `ppc64el`
- [x] RISC-V 64-bit `riscv64`

Update(s) Uploaded to Stable
----------------------------

**Primary Architectures**

- [ ] AMD64 `amd64`   
- [ ] AArch64 `arm64`

**Secondary Architectures**

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [ ] Loongson 3 `loongson3`
- [ ] PowerPC 64-bit (Little Endian) `ppc64el`
- [ ] RISC-V 64-bit `riscv64`

### CODE REVIEW

**Review by Nyovelt — COMMENTED** [review]
`amd64` build success

File: extra-ocaml/dune/autobuild/build [hunk 1]
```diff
@@ -1,21 +1,34 @@
 # Script adapted from Arch Linux community/dune.
+# https://github.com/archlinux/svntogit-community/blob/4cca5149bd5dd0c4211dac5e9be1c9bcd18b51dc/trunk/PKGBUILD
 
-abinfo "Bootstrapping Dune ..."
+## If $PKGDIR is placed under $SRCDIR, dune will search abdist and think the
+## installed executable as a dune file, and refuse to install.
+_SRCDIR="$SRCDIR"/dune-"$PKGVER"
+pushd $_SRCDIR
```
Comment (MingcongBai) [inline extra-ocaml/dune/autobuild/build]: Quote variable.

File: extra-ocaml/dune/autobuild/prepare [hunk 2]
```diff
@@ -0,0 +1,5 @@
+if [[ "${CROSS:-$ARCH}" = "loongson3" || \
```
Comment (MingcongBai) [inline extra-ocaml/dune/autobuild/prepare]: Use `if ab_match_arch loongson3 || ab_match_arch riscv64; then ...`.

File: extra-ocaml/findlib/autobuild/beyond [hunk 3]
```diff
@@ -0,0 +1,5 @@
+if [[ "${CROSS:-$ARCH}" != "loongson3" &&\
+      "${CROSS:-$ARCH}" != "riscv64" ]]
```
Comment (MingcongBai) [inline extra-ocaml/findlib/autobuild/beyond]: Ditto above.

File: extra-ocaml/findlib/autobuild/beyond [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+    abinfo "Fix file permissions ..."
+    chmod -x "$PKGDIR"/usr/lib/ocaml/findlib/findlib{,_top,_dynload}.a
```
Comment (MingcongBai) [inline extra-ocaml/findlib/autobuild/beyond]: Missing verbosity.

File: extra-ocaml/dune/autobuild/prepare [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+    "${CROSS:-$ARCH}" = "riscv64" ]]; then
+    abinfo "Disabling ELF filter and debug symbol stripper for ${CROSS:-$ARCH} ..."
+    export ABSTRIP=0 ABSPLITDBG=0
```
Comment (MingcongBai) [inline extra-ocaml/dune/autobuild/prepare:3]: Add a `#FIXME` comment for Autobuild3 bug.

File: extra-ocaml/ocaml-csexp/autobuild/prepare [hunk 6]
```diff
@@ -0,0 +1,5 @@
+if [[ "${CROSS:-$ARCH}" = "loongson3" || \
+    "${CROSS:-$ARCH}" = "riscv64" ]]; then
+    abinfo "Disabling ELF filter and debug symbol stripper for ${CROSS:-$ARCH} ..."
+    export ABSTRIP=0 ABSPLITDBG=0
```
Comment (MingcongBai) [inline extra-ocaml/ocaml-csexp/autobuild/prepare]: Ditto above for Autobuild3.

File: extra-ocaml/ocaml-pp/autobuild/build [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Using AB3's build/15-dune.sh will install unneeded integrated test component,
```
Comment (MingcongBai) [inline extra-ocaml/ocaml-pp/autobuild/build]: Doesn't make sense - review ABTYPE code to see what differences resulted in this issue.

File: extra-ocaml/ocaml-pp/autobuild/prepare [hunk 8]
```diff
@@ -0,0 +1,5 @@
+if [[ "${CROSS:-$ARCH}" = "loongson3" || \
+    "${CROSS:-$ARCH}" = "riscv64" ]]; then
+    abinfo "Disabling ELF filter and debug symbol stripper for ${CROSS:-$ARCH} ..."
+    export ABSTRIP=0 ABSPLITDBG=0
```
Comment (MingcongBai) [inline extra-ocaml/ocaml-pp/autobuild/prepare]: Ditto above.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

File: extra-ocaml/ocaml-pp/autobuild/build (hunk 7 above)
Comment (CamberLoid) [reply extra-ocaml/ocaml-pp/autobuild/build]: Thing is, `dune build` will build all available component, including those unneeded one for integrated testing purpose.

Comment (CamberLoid) [reply extra-ocaml/ocaml-pp/autobuild/build]: This is not a AB3 fault but a behavior of dune itself

**Review by MingcongBai — APPROVED** [review]

**Pull request merged**